core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java

```diff
@@ -100,6 +106,81 @@ public static EnumerableHashJoin create(
         condition, variablesSet, joinType);
   }
 
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      final RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null
+        || collation == RelCollations.EMPTY
+        || joinType == JoinRelType.FULL
+        || joinType == JoinRelType.RIGHT) {
+      return null;
+    }
+
+    List<Integer> requiredKeys = RelCollations.ordinals(collation);
+    ImmutableBitSet requiredKeySet = ImmutableBitSet.of(requiredKeys);
+    ImmutableBitSet leftKeySet = ImmutableBitSet.of(joinInfo.leftKeys);
+
+    // EnumerableHashJoin traits passdown shall only pass through collation to left input.
+    // It is because for EnumerableHashJoin always builds hash table on right input,
+    // and only non-hashed side can preserve ordering.
+
+    // Only consider exact key match for now
+    if (requiredKeySet.equals(leftKeySet)) {
+      RelTraitSet passthroughTraitSet = traitSet.replace(collation);
+      return Pair.of(passthroughTraitSet,
+          ImmutableList.of(
+              passthroughTraitSet,
+              passthroughTraitSet.replace(RelCollations.EMPTY)));
+    }
+
+    // TODO: consider super keyset and sub keyset
+    return null;
+  }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> deriveTraits(
+      final RelTraitSet childTraits, final int childId) {
+    // should only derive traits (limited to collation for now) from left join input.
+    if (childId > 0) {
```

Comment:
should be `assert childId == 0`, it is illegal to be greater than 0

Reply:
Ah yes. I think it is because this code block

```
        if (mode == DeriveMode.LEFT_FIRST
            || mode == DeriveMode.RIGHT_FIRST) {
          break;
        }
```